### PR TITLE
[camera_avfoundation] ignore audio samples until first video sample

### PR DESCRIPTION
**Pull request by misos1**
Samples from ios video and audio input should be regarded as arriving from independent streams because in fact they are going from independent sources. Video is going from instance `AVCaptureVideoDataOutput` while audio from `AVCaptureAudioDataOutput`. Therefore it is possible that one of them will be initialised soon or later than the other and the first few samples will be only audio samples so the video file will start without image data and will be black until the first video sample arrives. This patch addresses this by ignoring audio samples until the first video sample arrives (alternatively it could wait until both video and audio samples are actively arriving).

Fix https://github.com/flutter/flutter/issues/57831

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by hellohuanlin** [comment]
Feel free to request my review after you add some tests. In the mean time I am converting this to draft.

**Comment by stuartmorgan** [comment]
From triage: Ping @hellohuanlin on this review.

File: packages/camera/camera_avfoundation/CHANGELOG.md
Comment (hellohuanlin) [inline packages/camera/camera_avfoundation/CHANGELOG.md:3]: Ignores

File: packages/camera/camera_avfoundation/example/ios/RunnerTests/FLTCamSampleBufferTests.m
Comment (hellohuanlin) [inline packages/camera/camera_avfoundation/example/ios/RunnerTests/FLTCamSampleBufferTests.m:16]: let's avoid having a shared state across tests.

File: packages/camera/camera_avfoundation/example/ios/RunnerTests/FLTCamSampleBufferTests.m
Comment (hellohuanlin) [inline packages/camera/camera_avfoundation/example/ios/RunnerTests/FLTCamSampleBufferTests.m:102]: this test is unnecessarily complex. Can we just do a simpler version like this: 

didOutput: audio
didOutput: audio
didOutput: video

Then verify that the first 2 audio samples are omitted?

File: packages/camera/camera_avfoundation/example/ios/RunnerTests/FLTCamSampleBufferTests.m
Comment (misos1) [reply packages/camera/camera_avfoundation/example/ios/RunnerTests/FLTCamSampleBufferTests.m:49]: There is no other occurrence of `OCMClassMock([AVCaptureConnection class])` where it is assigned to a named variable. Naming is mixed among files either in form of `{something}Mock` or `mock{Something}` or even without word "mock" see for example `discoverySessionMock`, `handlerMock`, `mockDevice`, `videoSessionMock`. I cannot know which files are newer or contain preferred convention. So should I rather rewrite to use `mock{Something}` for that particular file?

File: packages/camera/camera_avfoundation/example/ios/RunnerTests/FLTCamSampleBufferTests.m
Comment (misos1) [reply packages/camera/camera_avfoundation/example/ios/RunnerTests/FLTCamSampleBufferTests.m:16]: `setUp` and `tearDown` are called before and after each individual test (function) so I do not understand how this is a shared state. This same thing is used also in `StreamingTest.m`. I also suppose there are going to be additional tests which will require a similar setup in near future.

File: packages/camera/camera_avfoundation/ios/Classes/FLTCam.m
Comment (misos1) [reply packages/camera/camera_avfoundation/ios/Classes/FLTCam.m:505]: In `FLTCam.m` it seems there is no way it could capture just audio. In the high level dart interface I see that just audio is optional with the `enableAudio` parameter in the `CameraController` constructor. Even if it recorded just audio it would probably result in black video with that audio as the output file format is always `AVFileTypeMPEG4` (this is just speculation but I never seen pure audio clip to be in mp4).

File: packages/camera/camera_avfoundation/example/ios/RunnerTests/FLTCamSampleBufferTests.m
Comment (misos1) [reply packages/camera/camera_avfoundation/example/ios/RunnerTests/FLTCamSampleBufferTests.m:102]: I would rather do audio, audio, video, audio to check whether not just the first sample is omitted but also that not all audio samples are omitted at all. You also mean to avoid that for loop?

File: packages/camera/camera_avfoundation/example/ios/RunnerTests/FLTCamSampleBufferTests.m
Comment (hellohuanlin) [reply packages/camera/camera_avfoundation/example/ios/RunnerTests/FLTCamSampleBufferTests.m:49]: hmmm, i probably remembered wrongly. I'd just follow the existing convention then, whatever it is.

File: packages/camera/camera_avfoundation/example/ios/RunnerTests/FLTCamSampleBufferTests.m
Comment (hellohuanlin) [reply packages/camera/camera_avfoundation/example/ios/RunnerTests/FLTCamSampleBufferTests.m:102]: yeah the for loop as well. keep the logic as simple as possible.

File: packages/camera/camera_avfoundation/example/ios/RunnerTests/FLTCamSampleBufferTests.m
Comment (hellohuanlin) [inline packages/camera/camera_avfoundation/example/ios/RunnerTests/FLTCamSampleBufferTests.m:41]: nit: something more descriptive, like `testDidOutputSampleBufferIgnoreAudioSamplesBeforeVideoSamples`?

File: packages/camera/camera_avfoundation/ios/Classes/FLTCam.m
Comment (hellohuanlin) [inline packages/camera/camera_avfoundation/ios/Classes/FLTCam.m:503]: can you also link the issue here in the comment?

**Review by hellohuanlin — APPROVED** [review]

**Review by hellohuanlin — APPROVED** [review]

**Comment by stuartmorgan** [comment]
@bparrishMines for second review

**Review by stuartmorgan — APPROVED** [review]
LGTM; once the version is re-bumped to resolve the conflicts with `main` we can land this. Thanks!

**Pull request merged**
